### ISSUE TITLE
Memset issue

### ISSUE THREAD

**Issue by ceesb** [NASA-SW-VnV/ikos#117]
Hi,

I am playing with some code that uses memset (or initializers like below) to set defaults in structures. I'm running into an issue there. This ikos_assert can not be proven:
```
#include <ikos/analyzer/intrinsic.h>
#include <stdint.h>

struct a {
	uint8_t b;
};


int main() {
	struct a i = {0};
	__ikos_assert(i.b == 0);
	return 0;
}
```
But this one can, the only difference being the value of the constant:
```
#include <ikos/analyzer/intrinsic.h>
#include <stdint.h>

struct a {
	uint8_t b;
};


int main() {
	struct a i = {1};
	__ikos_assert(i.b == 1);
	return 0;
}
```
Out of curiosity, I added the `--display-ar` flag to ikos to print the abstract representation. For the first example, I see the initializer is translated into `ar.memset`:

```
define si32 @main() {
#entry !entry successors={#1, #2} {
  {0: ui8}* $i = allocate {0: ui8}, 1, align 1
  si8* %1 = bitcast $i
  call @ar.memset(%1, 0, 1, 1, 0)
....
```

For the second example, I see `ar.memcpy` instead:
```
define si32 @main() {
#entry !entry successors={#1, #2} {
  {0: ui8}* $i = allocate {0: ui8}, 1, align 1
  si8* %1 = bitcast $i
  si8* %2 = ptrshift @__const.main.i, 1 * 0, 1 * 0
  call @ar.memcpy(%1, %2, 1, 1, 1, 0)
...
```

Any ideas as to what's happening? I may be able to fix if you point me into the right direction.

thx!
Cees

**Comment by arthaud** [comment]
Hi @ceesb,

Thanks for trying ikos.

This is an imprecision of the memory analysis. The current implementation abstracts the memory using cell variables. A cell variable is a symbolic variable that represents the value at a given memory location, offset, and size.

In the second example, there is a constant structure `__const.main.i` that has a global initializer. This creates the cell variable `Cell{__const.main.i, offset=0, size=1}` with a value of `1`. The memcpy then copies this cell into the local variable. This works fine.

In the first example, the local variable is initialized with memset. The current implementation of the abstract memset  (see [here](https://github.com/NASA-SW-VnV/ikos/blob/1c708ba5ba49870626f50ca020065916cb4ef06c/core/include/ikos/core/domain/memory/value.hpp#L1126)) only updates existing cell variables. It should also create one big cell variable equal to zero.

The reason why it is not currently implemented is because the current implementation does not know how to split cell variables, thus creating a big cell for the whole structure is pointless. If you have a structure with multiple fields, and you create a big cell of the size of the structure that is equal to zero, when you will read a specific field, the big cell will just be ignored, and the result will be unknown. This could be improved too, but we should be careful about endianness (see [here](https://github.com/NASA-SW-VnV/ikos/blob/1c708ba5ba49870626f50ca020065916cb4ef06c/core/include/ikos/core/domain/memory/value.hpp#L905) and [here](https://github.com/NASA-SW-VnV/ikos/blob/1c708ba5ba49870626f50ca020065916cb4ef06c/core/include/ikos/core/domain/memory/value.hpp#L587)).

In this case, fixing memset to create a big cell should be enough to fix the problem. Would you like to try to implement it?

**Comment by ceesb** [comment]
I see now that the memcpy/memset difference is introduced by LLVM. 

I tricked myself by incorrectly assuming that the "{1}" initializer would write 1's in the entire structure, but that's not true, and ikos rightly points out that the assertion below never holds ;-) Got to love C ..


```c
#include <ikos/analyzer/intrinsic.h>
#include <stdint.h>

struct a {
	uint8_t b;
	uint8_t c;
};


int main() {
	struct a i = {1};
	__ikos_assert(i.c == 1);
	return 0;
}
```

Just to understand the underlying issue with the memory cells you dscribe, I constructed another example using a struct with multiple elements of different sizes. The last assertion cannot be proven. This is caused by ikos not being able to split cells?

```c
#include <ikos/analyzer/intrinsic.h>
#include <stdint.h>

struct __attribute__((__packed__)) a {
	uint8_t a;
	uint8_t b;
	uint32_t c;
};

int main() {
	uint8_t buf[6] = {0xce,0xe5,0x01,0x00,0x00,0x00};

	struct a *i = (struct a*)&buf;
	__ikos_assert(i->b == 0xe5);
	__ikos_assert(i->c == 1);
	return 0;
}
```

**Comment by arthaud** [comment]
Yes, but in that case the problem is the opposite: it doesn't merge the cells.

**Comment by ceesb** [comment]
OK, thanks. I understand that the missing cell splitting and merging in ikos affects its capabilities on code that uses the case-to-struct trick on arrays of primitives types, and code that heavily depends on static initializers. 

I have another question. The target code I'm looking at is mixing pointers and primitive types. Ikos doesn't seem to appreciate this. The following example causes ikos to report it cannot bound the offset of a->b.

```c
#include <stdint.h>
#include <stdio.h>

struct a {
	uint8_t b;
};

struct b {
	struct a a;
	intptr_t p;
	uint8_t c;
};

void test(struct b *b) {
	struct a *a = NULL;

	a = (struct a*)(b->p);

	if(a->b) {
		printf("if branch\n");
	} else {
		printf("else branch\n");
	}
}

int main() {
	struct a a;
	a.b = 0;
	struct b b;
	b.p = (uintptr_t)&a;
	
	test(&b);
	
	return 0;
}

```

Is this because once you determine a variable to be a primitive type it may never be used as a pointer? A limitation of the memory model?

I hope you don't mind me peeking and poking a little bit, I'm just trying to understand how I can make the most use of ikos on some real-world code I'm reviewing.

**Comment by arthaud** [comment]
> OK, thanks. I understand that the missing cell splitting and merging in ikos affects its capabilities on code that uses the case-to-struct trick on arrays of primitives types, and code that heavily depends on static initializers.

Yes. So far, I didn't have any problem with it when analyzing safety critical code, but I imagine this can be a problem for some users.

> I have another question. The target code I'm looking at is mixing pointers and primitive types. Ikos doesn't seem to appreciate this. The following example causes ikos to report it cannot bound the offset of a->b.

The problem here is the cast from pointer to integer. When converting to an integer, we are losing all the relevant information on the pointer.

You can fix this by using a `void*` instead of `uintptr_t` in the structure.
Another trick would be to cast `b.p` to `struct a*` when using it:
```c
// In test
a = *((struct a**)(&b->p));
// In main
*((struct a**)(&b.p)) = &a;
```
But this is not easy to read..

> I hope you don't mind me peeking and poking a little bit, I'm just trying to understand how I can make the most use of ikos on some real-world code I'm reviewing.

I don't mind at all. I hope this can help other users as well.

**Comment by ivanperez-keera** [comment]
@ceesb I know it's been a few years, but I'm following up on this to see if you'd like to contribute this improvement for a future version of IKOS. Thanks!

**Comment by peckto** [comment]
Hi,

I came along this issue while debugging a false positive uninitialized variable finding in combination with structs and I thought this could be related to the split/merge cell topic.

Given the following c code:
```c
#include <stdio.h>
#include <string.h>

typedef struct
{
  int a;
  int b;
}
s_t;

s_t global_s1;
s_t global_s2 = {0,0};
s_t global_s3 = {2,2};
s_t global_s4;

void foo() {
    s_t local_s1;
    s_t local_s2;
    memset(&local_s1, 0, sizeof(local_s1));
    memset(&local_s2, 1, sizeof(local_s2));
    printf("%d\n", local_s1.b);     // OK
    printf("%d\n", local_s2.b);     // uninitialized??
}

void bar() {
    global_s1.a = 0;
    global_s2.a = 0;
    global_s3.a = 0;
    printf("%d\n", global_s1.b);    // uninitialized
    printf("%d\n", global_s2.b);    // uninitialized
    printf("%d\n", global_s3.b);    // OK
    printf("%d\n", global_s4.b);    // OK
    global_s4.a = 0;
    printf("%d\n", global_s4.b);    // OK
}

int main() {
    bar();
    foo();
    return 0;
}
```

IKOS reports the following findings:
```
# Results
main.c: In function 'foo':
main.c:22:5: warning: expression 'local_s2.b' might be uninitialized
    printf("%d\n", local_s2.b);     // uninitialized??
    ^
main.c: In function 'bar':
main.c:29:5: warning: expression 'global_s1.b' might be uninitialized
    printf("%d\n", global_s1.b);    // uninitialized
    ^
main.c: In function 'bar':
main.c:30:5: warning: expression 'global_s2.b' might be uninitialized
    printf("%d\n", global_s2.b);    // uninitialized
    ^
```

The problem seems to occur when the memory is initialized with `memset`.
This sounds like what you mentioned earlier
> ... the local variable is initialized with memset. The current implementation of the abstract memset (see [here](https://github.com/NASA-SW-VnV/ikos/blob/1c708ba5ba49870626f50ca020065916cb4ef06c/core/include/ikos/core/domain/memory/value.hpp#L1126)) only updates existing cell variables. It should also create one big cell variable equal to zero.

What surprises me is, that line 22 is also reported, the local variable initialized with `memset` to `1`, instead of `0`.
Maybe this is an unrelated issue?
